Feature/out of band nonsms

achen2401 opened the pull request:
https://www.pivotaltracker.com/story/show/184554218
Allow out of band non-SMS messages to be entered and saved.

Changes include:

- new UI for entering non-SMS messages (see screenshot)
   options to be chosen from are `Issac send`, `Enter manual message`, `Enter comment`
<img width="682" alt="Screen Shot 2023-04-18 at 10 33 09 AM" src="https://user-images.githubusercontent.com/12942714/232810356-a3183a9b-b133-4e3a-b76c-56e386524584.png">

mobile view:
![mobileView](https://user-images.githubusercontent.com/12942714/232873126-6eb4fee6-36f0-4bf0-95b5-58e759017516.png)



- When `Enter manual message` is chosen, screenshot of the UI:
<img width="663" alt="Screen Shot 2023-04-18 at 10 33 59 AM" src="https://user-images.githubusercontent.com/12942714/232810569-64816be1-699d-40b9-b74f-210cc24e8ce9.png">


- When `Enter comment` is chosen, screenshot of the UI:
<img width="670" alt="Screen Shot 2023-04-18 at 10 34 38 AM" src="https://user-images.githubusercontent.com/12942714/232810709-72a09a62-6720-47dd-8803-3612cce10f15.png">



- Screenshot for different types of messages in the messages panel:
<img width="673" alt="Screen Shot 2023-04-18 at 11 06 41 AM" src="https://user-images.githubusercontent.com/12942714/232820593-c4c4d558-b7e8-46e7-abbf-0b5dce3f7f58.png">


non-sms messages and comments are made distinct from ISACC sent messages with different sticky-note like style


- when clicking on the `I` information icon, the following popup will appear:
<img width="692" alt="Screen Shot 2023-04-17 at 3 12 58 PM" src="https://user-images.githubusercontent.com/12942714/232587770-91861eea-45b1-489a-91e0-3ced0177a0f5.png">



- the Non SMS message will be saved as a FHIR resource of type `Communication`:
Example:
```
{
        "resourceType": "Communication",
        "id": "1509244",
        "meta": {
          "versionId": "1",
          "lastUpdated": "2023-04-10T12:48:42.200-04:00"
        },
        "partOf": [
          {
            "reference": "CarePlan/1509228"
          }
        ],
        "category": [
          {
            "coding": [
              {
                "system": "https://isacc.app/CodeSystem/communication-type",
                "code": "isacc-non-sms-message"
              }
            ]
          }
        ],
        "subject": {
          "reference": "Patient/2cda5aad-e409-4070-9a15-e1c35c46ed5a"
        },
        "sent": "2023-04-10T16:48:40.560Z",
        "payload": [
          {
            "contentString": "test"
          }
        ],
        "note": [
          {
            "text": "Non-SMS comment message, staff entered"
          }
        ]
 }
```
**sent** or **received** element is populated via date/time entered/selected via datetimepicker
**payload** element is populated the text entered for the content
**note** element  is populated by text that indicates this is a Non-SMS communication, e.g. `Non-SMS sent manual message, staff entered`
The resource is marked as "non-sms" via the `category`, e.g. ` "category": [
    {
      "coding": [
        {
          "system": "https://isacc.app/CodeSystem/communication-type",
          "code": "isacc-non-sms-message"
        }
      ]
    }
  ]` element

Commit:
Non-SMS UI - first commit

Commit:
add info dialog

Commit:
remove un-instended changes

Commit:
remove un-intended changes

Commit:
changes based on feedback

Commit:
syntax fix

Commit:
fix code syntax

Commit:
Merge branch 'dev' of https://github.com/uwcirg/isacc-messaging-client-sof into feature/out-of-band-nonsms

Commit:
remove commented out code

mcjustin commented:
The "comment" capacity will partially address https://www.pivotaltracker.com/story/show/184772024 (which isn't required for May's release) - it may even be sufficient.

mcjustin commented:
@achen2401  please use the verbiage from Kate: 
    "reply" (from patient)
    "response" (from system/provider)

The screenshot is confusing to me, it looks like there are patient replies on both the left and right. https://user-images.githubusercontent.com/12942714/230949376-3d07928c-7891-4db8-b1a0-093f16b65c97.png

Commit:
fix based on feedback

Commit:
text, formatting changes

Commit:
bug fix

Commit:
add feedback

Commit:
bug fix

Commit:
Merge branch 'dev' of https://github.com/uwcirg/isacc-messaging-client-sof into feature/out-of-band-nonsms

Commit:
minor attribute fix

achen2401 commented:
@hannahburkhardt Will you review my changes when you have a moment?

Commit:
styling fix

Commit:
minor styling fix

Commit:
bug fix

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+            minHeight: "40px"
+          }}
+        >
+          <Tab value="sms" label="Issac send" {...tabProps}/>
```
```suggestion
          <Tab value="sms" label="ISACC send" {...tabProps}/>
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+        <TextField
+          multiline
+          value={this.state.activeMessage?.content ?? ""}
+          placeholder={"Type message for ISSAC to send"}
```
Two Cs, as in *C*aring *C*ontacts :)
```suggestion
          placeholder={"Type message for ISACC to send"}
```

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      : false;
+
+    if (isNonSmSMessage) {
+      incoming = message.received ? true : false;
```
```suggestion
      incoming = !!message.received;
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      )
+    )
+      ? true
+      : false;
```
```suggestion
    const isNonSmSMessage = !!message.category.find((c: ICodeableConcept) =>
      c.coding.find((coding: ICoding) =>
        IsaccMessageCategory.isaccNonSMSMessage.equals(coding)
      )
    );
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      </Typography>
+    ) : (
+      false
+    );
```
```suggestion
    ) : false;
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+  async getCommunications(
+    client: Client,
+    carePlanId: string
+  ): Promise<Communication[]> {
```
I'm wondering what code style you're using? I don't generally break up method signatures unless the line is over 80 columns (characters) wide. Either works for me, but we should be consistent!

achen2401 replied:
I am using [VS code eslint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) extension to format my code.  I will probably need to tweak its setting.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+            incoming
+              ? "reply (from recipient)"
+              : message.sent
+              ? "response (from provider)"
```
```suggestion
              ? "response (from author)"
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+    return await client
+      .request({
+        url: `/Communication?${params}`,
+        headers: {
+          "Cache-Control": "no-cache",
+        },
```
Should this always be added everywhere? If so, is there a way to refactor to prevent code duplication?

achen2401 replied:
I think so but it seems out of the scope for this PR.  I added a TODO in the code so we can come back to address this.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+    // Communication?part-of=CarePlan/${carePlanId}
+    let params = new URLSearchParams({
+      "part-of": `CarePlan/${carePlanId}`,
+      // _sort: "-sent",
```
Can this line be removed (rather than commented out)?

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
-        return this._alignedRow(incoming, msg, timestamp, bubbleStyle, priority, index, themes);
-
+    let messageBoxProps = {
+      maxHeight: 548,
```
Where does this oddly specific number come from? Can we somehow avoid injecting "magic numbers" in this way?

achen2401 replied:
It was originally [600](https://github.com/uwcirg/isacc-messaging-client-sof/blob/main/src/components/MessagingView.tsx#L48)  I reduced it to try to fit the messages list and the bottom widget for entering message to fit within the viewport.  I have since changed it back to 600, as you are right, it is an arbitrary number and there is probably a better way to specify this.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      });
+    const tabProps = {
+        sx: {
+            fontSize: "0.85rem",
```
Is there a standard style from the Material UI kit that we could use here instead, so we don't have to inject this magic number?

achen2401 replied:
We could probably do this via `GlobalStyles` ([here](https://mui.com/material-ui/customization/how-to-customize/#3-global-theme-overrides)) for a consistent override I have removed the magic number for now.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+            <Typography variant="body1">
+              <p>
+                <b>ISACC send</b> are messages sent or received directly by
+                ISACC, which include texts, “sms”, iMessages, etc.
```
```suggestion
                ISACC, including text messages (SMS), iMessages, etc.
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+              <p>
+                <b>Manual messages</b> mean messages external to ISACC,
+                including manually entered notifications of emails, phone calls,
+                postal mail, or other non-automated communication.
```
```suggestion
                <b>Manual messages</b> are manually entered records of communications that took place outside of ISACC, such as emails, phone calls,
                postal mail, or other non-automated communications.
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+          <DialogContent>
+            <Typography variant="body1">
+              <p>
+                <b>ISACC send</b> are messages sent or received directly by
```
```suggestion
                <b>ISACC messages</b> are sent or received directly by
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+                postal mail, or other non-automated communication.
+              </p>
+              <p>
+                <b>Comments</b> are contact/note on the recipient.
```
What is meant by "contact/note"? Is it a contact (I thought those are "manual messages")? Can we just remove the explanation, I think people know what a "comment" is and they can use the field to enter information in any way they see fit.

achen2401 replied:
addressed. thanks.  I removed the explanation.

hannahburkhardt committed:
fix typo succes->success

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+            at
+          </Typography>
+          <DateTimePicker
+            label="date & time"
```
```suggestion
            label="Date & time"
```

achen2401 replied:
addressed. thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      this.state.activeMessage?.status === "received"
+        ? this.state.activeMessage.date
+        : null;
+    const noteAboutCommunication = `Non-SMS ${this.state.activeMessage?.type} message, staff entered`;
```
1. spelling fix
```suggestion
    const noteAboutCommunication = `Non-SMS ${this.state.activeMessage?.type} message, staff-entered`;
```
2. in what sense is this a 'message' if it is a comment? This is misleading. Users will think these are messages exchanged with the patient

achen2401 replied:
addressed. thanks.  I removed the text `message`.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      <>
+        <Stack
+          direction={{
+            xs: "column",
```
Is it really necessary to stack these on top of each other in the phone-sized view? It makes the screen very tall vertically, and will result in users having to scroll. It looks to me like there should be enough space to keep the radio button selector and date text field next to each other.
<img width="473" alt="image" src="https://user-images.githubusercontent.com/4624654/232350197-ab0fa0f4-2e36-4eb1-a168-29f1d78d1c53.png">

achen2401 replied:
addressed.  thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+      false
+    );
+
+    const comment: React.ReactNode = isNonSmSMessage ? (
```
Can you find a better variable name? Maybe "label"? This is confusing because comment is a type of message that can be displayed here

achen2401 replied:
renamed.  thanks.

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+              ? "response (from provider)"
+              : ""
+          }`}
+        </Typography>
```
This is redundantly stating something that's already conveyed via other UI elements, and just clutters the UI

achen2401 replied:
I added this change in response to Justin's comment [here](https://github.com/uwcirg/isacc-messaging-client-sof/pull/18#issuecomment-1502215761)

hannahburkhardt commented on src/components/MessagingView.tsx:
```diff
@@ -4,333 +4,943 @@ import {FhirClientContext, FhirClientContextType} from "../FhirClientContext";
+    index: number,
+    themes: string[],
+    comment: React.ReactNode
+  ) {
```
The intention with the `_alignedRow` method was to pass in primitive types only, and then turn those into a component for rendering. Now, all are primitives except one (comment) which is already a component. If we want to change what the final component looks like, we now have to touch 2 different methods. It might be good to refactor this so that "comment" is a string, and that is rendered into the ReactNode only within this method. Something like this:

```suggestion

    const comment: React.ReactNode = isNonSmSMessage ? [note, `${
            message.sent ? "sent" : "received"
          } on ${MessagingView.displayDateTime(
            message.sent ? message.sent : message.received
          )}`].join("\n") : [note, `${
            incoming
              ? "reply (from recipient)"
              : message.sent
              ? "response (from provider)"
              : ""
          }`].join('\n');

    return this._alignedRow(
      incoming,
      msg,
      timestamp,
      bubbleStyle,
      priority,
      index,
      themes,
      comment
    );
  }

  private _alignedRow(
    incoming: boolean,
    message: string,
    timestamp: string,
    bubbleStyle: object,
    priority: string,
    index: number,
    themes: string[],
    comment: string
  ) {
  const commentDisplay = <Typography variant="caption" color="text.secondary">
          {comment}
        </Typography>
```

achen2401 replied:
thanks.  I used your suggested code here.

hannahburkhardt commented:
I'm not sure the speech bubble design is a good choice for this kind of message. It really gives the impression that these are messages that were exchanged, verbatim as shown, between author and recipient. While it's possible that staff just entered the text of an email exactly as it was written by a recipient, this is not at all clear and I think this isn't really what would happen. For example, a patient might email "I'm okay" at 12pm. If staff go in and enter this, they might write "They're okay", which will easily be misinterpreted by a reader as the patient having said "They're okay" (their dogs, their children, ...). It's not clear what was said because it's displayed as if it's a verbatim message but it is not. I get that we're trying to use color, labels, and info messages to prevent this interpretation, but I think we can just use a different design that doesn't give that impression in the first place. Maybe something like this?
<img width="557" alt="image" src="https://user-images.githubusercontent.com/4624654/232355977-121c2883-2e88-432a-80ce-488fb9dfa872.png">

The only reason we include these items in the message card is because we want them all on the same timeline - so we can use a different display style for these, it doesn't have to be speech bubbles (which very clearly convey "verbatim speech"). I also think the yellow color gives the impression that there's something wrong (yellow is a warning color).

Also, are "manual messages" and "comments" supposed to look different somehow?

hannahburkhardt committed:
variable case fix

hannahburkhardt commented:
Can we come up with shorter labels for the tabbed entry component? This becomes to big to fit on screen very quicky. On a phone, you definitely couldn't see all 3 tabs.
<img width="426" alt="image" src="https://user-images.githubusercontent.com/4624654/232356837-28c11cbc-0ec1-46a7-b2a7-749c0a05f56d.png">

I also think that "manual message" will probably not be readily understood. I know we have the explanation written under the info icon but maybe we can do better. How about "ISACC message" vs "Out-of-band message"? Out of band is a new concept, which can be learned quickly, and will not overlap with ideas that users already have in their heads, which might cause confusion.

hannahburkhardt commented:
Are the corner radii of these text boxes different on purpose?
<img width="590" alt="image" src="https://user-images.githubusercontent.com/4624654/232356946-3c2bf8eb-278b-4a86-8e8e-9d922c09a611.png">

<img width="577" alt="image" src="https://user-images.githubusercontent.com/4624654/232356990-3d9ea749-90c0-4926-a84f-8c4d5cb3955d.png">

hannahburkhardt commented:
I understand why you added the gray background here (to set apart the selectors, I assume), but it makes the view look a bit fragmented now. Also, the gray background has sharp corners. I wonder if there's anything we can do to make this look a bit cleaner? Maybe we can do without the gray background?
<img width="587" alt="image" src="https://user-images.githubusercontent.com/4624654/232357468-d5e1ede5-e1d2-4203-a234-2ae1efc29e5b.png">

hannahburkhardt commented:
I'm not sure if I did something to break this, because it looks fine on femr, but the message component seems too wide for the screen while the patient info component is not quite wide enough...? 🤔
<img width="469" alt="image" src="https://user-images.githubusercontent.com/4624654/232357622-213d3375-4fb6-4971-88e6-2d629f5ea51c.png">

hannahburkhardt commented:
Comments appear left-aligned or right-aligned depending on whether I selected "sent" or "received" on the "enter manual message" tab. If comments can be "sent" or "received", we should show the radio button selector on this tab as well. If not, comments should not be left or right aligned (maybe centered?)

<img width="569" alt="image" src="https://user-images.githubusercontent.com/4624654/232360499-8427deef-35c4-4ec6-bc62-c71d97203536.png">

Also, "Comment message" is confusing terminology.. This isn't really a message in any sense of the word, so we should probably just write comment. 
But, I'm a big fan of using self-explanatory styles rather than labeling things. I think it's better to use a different style for non-SMS messages than SMS messages, rather than adding a label "non-sms manual message". See my other comment

achen2401 committed:
Update src/components/MessagingView.tsx

Co-authored-by: Hannah <ha.al.bu@gmail.com>

Commit:
changes based on feedback

Commit:
fix UI based on feedback

Commit:
fix text display

achen2401 commented:
@hannahburkhardt I believe I addressed all the code changes you requested.  And thank you for your helpful comments about the UI.  I incorporated some of your suggestions, i.e. removed gray background that was used to set apart from the message type selector, make the non-sms messages/comments distinct from those sent from ISACC and attached the latest screenshots for the UI changes.   There were things you suggested, such as `shorter labels for the tabbed entry component`, `renaming manual messages to make it less confusing` -  that are very helpful but I am not certain what the best changes are for those.   I may need help from Sie-ce on refining the styles and I think it'd be good to hear input from Bill, Justin and Kate and other stake holders once these UI changes are available for them to review.

Commit:
Merge branch 'dev' of https://github.com/uwcirg/isacc-messaging-client-sof into feature/out-of-band-nonsms

Commit:
fix style

Commit:
fix style

Commit:
fix note style

Commit:
Merge branch 'dev' of https://github.com/uwcirg/isacc-messaging-client-sof into feature/out-of-band-nonsms

mcjustin reviewed (approved):
@achen2401  per our slack chat, merge when ready. I'd prefer to get this out there for our many stakeholders/designers to play with and refine it later in a separate PR, but I'll defer to you.

hannahburkhardt commented:
@achen2401  I love what you did with the manual entry/notes display appearance - looks like a bit like a post-it now, but cleaner. Perfect!!
